File: C/Zula.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void driverLogin();
void driverSignUp();
void driverPortal();
void customerLogin();
void customerSignUp();
void customerPortal();
void AdminPortal();

typedef struct
{
    char name[30], passwd[30];
    int age, cabId;
} Driver;

typedef struct
{
    char name[30], passwd[30];
    int age, custId;
} Customer;

typedef struct
{
    Driver *driver;
    Customer *customer;
    size_t used;
    size_t size;
} Array;

Array *Drivers;
Array *Customers;

void initDriver(Array *a, size_t initialSize)
{
    a->driver = (Driver *)malloc(initialSize * sizeof(Driver));
    a->used = 0;
    a->size = initialSize;
}

void initCustomer(Array *a, size_t initialSize)
{
    a->customer = (Customer *)malloc(initialSize * sizeof(Customer));
    a->used = 0;
    a->size = initialSize;
}

void insertDriver(Array *a, Driver *element)
{
    if (a->used == a->size)
    {
        a->size += 1;
        a->driver = (Driver *)realloc(a->driver, a->size * sizeof(Driver));
    }
    a->driver[a->used++] = *element;
}

void insertCustomer(Array *a, Customer *element)
{
    if (a->used == a->size)
    {
        a->size += 1;
        a->customer = (Customer *)realloc(a->customer, a->size * sizeof(Customer));
    }
    a->customer[a->used++] = *element;
}

void freeDriver(Array *a)
{
    free(a->driver);
    a->driver = NULL;
    a->used = a->size = 0;
    free(a);
}

void freeCustomer(Array *a)
{
    free(a->customer);
    a->customer = NULL;
    a->used = a->size = 0;
    free(a);
}

void mainMenu()
{
    int option;
    printf("WELCOME TO THE ZULA !! ");
    printf("\n1.Driver Portal \n2.Customer Portal \n3.Admin Portal \n4.Exit\n");
    printf("\nChoose option : ");
    scanf("%d", &option);
    printf("This is the option %d\n", option);
    switch (option)
    {
    case 1:
        system("clear");
        printf("Driver Portal");
        driverPortal();
        break;
    case 2:
        system("clear");
        printf("Customer Portal ");
        customerPortal();
        break;
    case 3:
        system("clear");
        printf("Admin Portal ");
        AdminPortal();

        break;
    case 4:
        system("clear");
        printf("Exiting...\n");
        exit(0);
        break;
    default:
        system("clear");
        printf(" Invalid ");
        mainMenu();
    }
}

void AdminPortal()
{
    int choice;
    printf("\n1.Show Driver Records \n2.Show Customer Records \n3.Main Menu");
    printf("\nChoose option : ");
    scanf("%d", &choice);
    switch (choice)
    {
    case 1:
        system("clear");
        printf("\nDriver Records");

        printf("\nCabId\tName\tAge");
        printf("\n-------------------");
        for (int i = 0; i < Drivers->size; i++)
        {
            printf("\n%d\t%s\t%d", Drivers->driver[i].cabId, Drivers->driver[i].name, Drivers->driver[i].age);
        }
        printf("\n");
        AdminPortal();
        break;
    case 2:
        system("clear");
        printf("\nCustomer Records");

        printf("\nCustId\tName\tAge");
        printf("\n-------------------");
        for (int i = 0; i < Customers->size; i++)
        {
            printf("\n%d\t%s\t%d", Customers->customer[i].custId, Customers->customer[i].name, Customers->customer[i].age);
        }
        printf("\n");
        AdminPortal();
        break;

    case 3:
        system("clear");
        mainMenu();
        break;
    default:
        system("clear");
        printf("Invalid");
        AdminPortal();
    }
}

void driverPortal()
{
    int choice;
    Driver newDriver;
    printf("\n1.Login \n2.SignUp \n3.Main Menu\n");
    printf("\nEnter your choice: ");
    scanf("%d", &choice);
    switch (choice)
    {
    case 1:
        driverLogin();
        break;
    case 2:
        driverSignUp();
        break;
    case 3:
        system("clear");
        mainMenu();
        break;
    default:
        printf("\nInvalid Choice");
        driverPortal();
    }
}

void driverLogin()
{
    char driverName[30], driverPasswd[30];
    int flag = 0;
    printf("\nEnter Driver Name: ");
    scanf("%s", driverName);
    for (int i = 0; i < Drivers->size; i++)
    {
        if (strcmp(Drivers->driver[i].name, driverName) == 0)
        {
            flag = 1;
            printf("\nEnter Driver Password: ");
            scanf("%s", driverPasswd);
            if (strcmp(Drivers->driver[i].passwd, driverPasswd) == 0)
            {
                system("clear");
                printf("\nDriver Login Successful");
                driverPortal();
            }
            else
            {
                system("clear");
                printf("\n Password Error");
                driverPortal();
            }
        }
    }
    if (flag == 0)
    {
        system("clear");
        printf("\nDriver Not Found");
        driverPortal();
    }
}

void driverSignUp()
{
    Driver newDriver;
    printf("\nEnter Driver Name: ");
    scanf("%s", newDriver.name);
    printf("\nEnter Driver Password: ");
    scanf("%s", newDriver.passwd);
    printf("\nEnter Driver Age: ");
    scanf("%d", &newDriver.age);
    newDriver.cabId = Drivers->used + 1;
    insertDriver(Drivers, &newDriver);
    system("clear");
    printf("\nDriver SignUp Successful");
    driverPortal();
}

void customerPortal()
{
    int choice;
    printf("\n1.Login \n2.SignUp \n3.Main Menu\n");
    printf("\nEnter your choice: ");
    scanf("%d", &choice);
    switch (choice)
    {
    case 1:
        customerLogin();
        break;
    case 2:
        customerSignUp();
        break;
    case 3:
        system("clear");
        mainMenu();
        break;
    default:
        printf("\nInvalid Choice");
        customerPortal();
    }
}

void customerLogin()
{
    char customerName[30], customerPasswd[30];
    int flag = 0;
    printf("\nEnter Customer Name: ");
    scanf("%s", customerName);
    for (int i = 0; i < Customers->size; i++)
    {
        if (strcmp(Customers->customer[i].name, customerName) == 0)
        {
            flag = 1;
            printf("\nEnter Customer Password: ");
            scanf("%s", customerPasswd);
            if (strcmp(Customers->customer[i].passwd, customerPasswd) == 0)
            {
                system("clear");
                printf("\nCustomer Login Successful");
                customerPortal();
            }
            else
            {
                system("clear");
                printf("\n Password Error");
                customerPortal();
            }
        }
    }
    if (flag == 0)
    {
        system("clear");
        printf("\nCustomer Not Found");
        customerPortal();
    }
}

void customerSignUp()
{
    Customer newCustomer;
    printf("\nEnter Customer Name: ");
    scanf("%s", newCustomer.name);
    printf("\nEnter Customer Password: ");
    scanf("%s", newCustomer.passwd);
    printf("\nEnter Customer Age: ");
    scanf("%d", &newCustomer.age);
    newCustomer.custId = Customers->used + 1;
    insertCustomer(Customers, &newCustomer);
    system("clear");
    printf("\nCustomer SignUp Successful");
    customerPortal();
}

int main()
{
    system("clear");
    Drivers = (Array *)malloc(sizeof(Array));
    Customers = (Array *)malloc(sizeof(Array));
    initDriver(Drivers, 3);
    insertDriver(Drivers, &(Driver){.name = "John", .age = 20, .cabId = 1, .passwd = "123"});
    insertDriver(Drivers, &(Driver){.name = "Jane", .age = 21, .cabId = 2, .passwd = "123"});
    insertDriver(Drivers, &(Driver){.name = "Jack", .age = 22, .cabId = 3, .passwd = "123"});
    initCustomer(Customers, 3);
    insertCustomer(Customers, &(Customer){.name = "Alpha", .passwd = "111", .age = 20, .custId = 1});
    insertCustomer(Customers, &(Customer){.name = "Beta", .passwd = "111", .age = 21, .custId = 2});
    insertCustomer(Customers, &(Customer){.name = "Gamma", .passwd = "111", .age = 22, .custId = 3});

    mainMenu();

    freeDriver(Drivers);
    freeCustomer(Customers);
    return 0;
}

File: Python/Zula.py
#Locations 
Locations = {'A':0,'C':4,'D':7,'F':9,'B':15,'G':18,'H':20,'E':23}
class Database:
    def __init__(self,name,data):
        self.name = name
        self.data = data

    def add(self,data):
        self.data.append(data)
    
    def user (self,user):
        for data in self.data:
            if data.name == user:
                return data
        
    def info(self):
        info = {}
        for i in self.data :
            info [i.name]= i.pswd
        return info
    
    def iD(self,id):
        for i in self.data:
            if i.id == id:
                return i
    

class Customer:
    class Rides:
        def __init__(self,cabId,source,destination):
            self.cabId = cabId
            self.source =source
            self.destination=destination
            self.fare = abs(Locations[destination] - Locations[source]) *10
        
        def rideDetails(self):
            print("{:<7} {:<13} {:<13} {:<9} ".format(self.source,self.destination,self.cabId,self.fare))
    id = 1


    def __init__(self,name,pswd,age,rides=[]):
        self.id = Customer.id
        self.name = name
        self.pswd = pswd
        self.age = age 
        self.rides = rides
        Customer.id +=1

    def details(self):
        print(f'{self.id}   {self.name}    {self.pswd}    {self.age}')
    
    def rideAdd(self,cabId,source,destination):
        self.rides.append(self.Rides(cabId,source,destination))
    
    def rideHistory(self):
        if self.rides:
            print ("{:<7} {:<13} {:<13} {:<5} ".format('Source','Destination','Cab-ID','Fare'))
            for i in self.rides:
                i.rideDetails()
        else:
            print('\nNo Records Found')






class Drivers:
    class Trips:
        def __init__(self,customerId,source,destination):
            self.customerId = customerId
            self.source =source
            self.destination=destination
            self.fare = abs(Locations[destination] - Locations[source]) *10
            self.ZCommission = self.fare * 0.3
        
        def tripDetails(self):
            print("{:<7} {:<13} {:<13} {:<9} {:<5}".format(self.source,self.destination,self.customerId,self.fare,self.ZCommission))

            
    id = 1
    def __init__(self,name,pswd,age,location,trips = []):
        self.id = Drivers.id
        self.name = name
        self.pswd = pswd
        self.age = age 
        self.location = location
        self.trips =[]
        if trips:
            for trip in trips:
                self.trips.append(self.Trips( *trip))
        Drivers.id +=1
    
    def details(self):
        if self.id == 1:
            print ("{:<5} {:<5} {:<10} {:<5} {:<5}".format('ID','Name','Password','Age','Location'))
        print ("{:<5} {:<5} {:<10} {:<7} {:<5}".format(self.id,self.name,self.pswd,self.age,self.location))
    
    def tripsInfo(self):
        if self.trips:
            print ("{:<7} {:<13} {:<13} {:<5} {:<5}".format('Source','Destination','Customer-ID','Fare','ZULA Commission'))
            for i in self.trips:
                i.tripDetails()
        else:
            print('\nNo Records Found')
    
    def tripadd(self,source,destination,customerId):
        self.trips.append(self.Trips(customerId,source,destination))

#Customers
ww = Customer('ww', 55,25)
xx = Customer('xx', 66,36)
yy = Customer('yy', 77,31)

#Drivers
aaa= Drivers('aaa',111,25,'D',[(4,'D','H'),(2,'E','G'),(2,'C','B')])
bbb=Drivers('bbb',222,36,'G')
ccc =Drivers('ccc',333,31,'H')
ddd= Drivers('ddd',444,28,'A')


CustomerDB =Database('Customer',[ww,xx,yy])
DriversDB = Database('Drivers',[aaa,bbb,ccc,ddd])


def clrscr():
    import os 
    os.system('cls')


def driver_Portal(username = None , password =None):
    def derror():
        print('1.Driver Login \n2.Main Menu ')
        choice = int(input("Choose option : "))
        if choice ==1:
            driver_Portal()
        elif choice == 2:
            clrscr()
            main()
            quit()
        else:
            clrscr()
            print('Enter Appropriate value')
            main()
            quit()

    if username==None and password == None:
        username = input("\nEnter the username: ")
        password = int(input("Enter the password: " ))
    if username in DriversDB.info().keys():
        if password == DriversDB.info()[username]:
            clrscr()
            print('Driver Logged In \n')
            opt = int(input("1.Show Records \n2.Main Menu \n3.Exit \nChoose Option: "))
            if opt ==1:
                DriversDB.user(username).tripsInfo()
                input()
                driver_Portal(username,password)
            elif opt == 2:
                main()
                quit()
            else:
                quit()
        else:
            clrscr()
            print("Invalid password")
            derror()
            quit()

    else:
        clrscr()
        print('Invalid Username !!')
        derror()
        quit()

def customer_Portal():
    def nearby_Drivers(source):
        l=[]
        for i in DriversDB.data:
            l.append(abs(Locations[source]-Locations[i.location]))
        v = l.index(min(l))+1
        return v


    def customer_Login(username= None,password=None): #Customer Login
        if username == None:
            username = input("\nEnter the username: ")
            password = int(input("Enter the password: " ))
        if username in CustomerDB.info().keys():
            if password == CustomerDB.info()[username]:
                clrscr()
                print("1.Book ride \n2.Ride History  \n3.Main Menu")
                Options = int(input("Choose an option : "))
                if (Options == 1):
                    print("Locations : ", end=' ')
                    for i in Locations:
                        print(i,end = ' ')
                    source = input("\nEnter current location : ")
                    destination = input("Enter drop location : ")
                    if (source == destination):
                        clrscr()
                        print("Source and destination shouldn't be same")
                        customer_Login(username,password)
                        quit()
                    Fare = 10*(Locations[destination]-Locations[source])
                    print("Total Fare Price : ",abs(Fare))
                    cnf = int(input('Confirm Ride (1 - YES/0 - NO)'))
                    if cnf:
                        cab = nearby_Drivers(source)
                        print("Booking confirmed - Cab id ",cab)
                        DriversDB.iD(cab).tripadd(source,destination,CustomerDB.user(username).id)
                        DriversDB.iD(cab).location = destination
                        CustomerDB.user(username).rideAdd(cab,source,destination)

                        print("1.Main menu \n2.Exit")
                        cho = int(input('Choose Option : '))
                        if cho == 1:
                            main()
                        else:
                            print("Exiting")
                            quit()
                    else:
                        customer_Login(username,password)

                elif Options ==2:
                    CustomerDB.user(username).rideHistory()
                    input()
                    customer_Login(username,password)
                elif Options == 3:
                    main()
                    quit()
                else:
                    clrscr()
                    print("Error Occured Sign in ")
                    customer_Login()
            else:
                clrscr()
                print("Password Error")
                customer_Login()
                quit()
        else:
            clrscr()
            print("User doesn't exist")
            customer_Login()
    
    def customer_Signup():
        username = input("Enter the username :")
        if username in CustomerDB.info().keys():
            clrscr()
            print("Username already taken. Try Different Username")
            customer_Signup()
        else:
            password = int(input("Enter numerical passowrd: "))
            age = int(input("Enter your age : "))
            CustomerDB.add(Customer(username,password,age))
            clrscr()
            print("User Created")
            customer_Login()
            quit()
    
    print("1.Login \n2.Signup \n3.Exit")
    option= int(input("Choose an option: "))
    if option ==1:
        customer_Login()
        quit()
    elif option == 2:
        customer_Signup()
        quit()
    elif option == 3:
        quit()
    else:
        clrscr()
        print('Enter appropriate Value')
        customer_Login()

def admin_Portal():
    id = int(input("Enter Cab ID :"))
    dr = DriversDB.iD(id)
    print("Total Number of Trips :", len(dr.trips))
    totalF =0
    zC = 0
    for i in dr.trips:
        totalF = totalF + i.fare
        zC = zC + i.ZCommission
    print("Total Fare Collected :", totalF)
    print("Total ZULA Commission :", zC)
    print('Trip Details :')
    dr.tripsInfo()
    
    opt = int(input("\n1.Main Menu \n2.Exit \nChoose a option : "))
    if opt == 1:
        clrscr()
        main()
        quit()
    elif opt == 2:
        quit()
    else:
        print("Error Retry")
        admin_Portal()

def main():
    print("Welcome to ZULA!!")
    print("1.Cab driver login\n2.Customer login\n3.Admistration login\n4.Quit ")
    service = int(input("Please choose service : "))
    if service == 1:
        clrscr()
        driver_Portal()
    elif service == 2:
        clrscr()
        customer_Portal()
    elif service == 3:
        clrscr()
        admin_Portal()
    elif service ==4:
        print(" Exiting ") 
        quit() 
    else:
        main()
        quit()


if __name__ =='__main__':
    clrscr()
    main()